e2e/testcases/namespace_selectors_test.go

```diff
@@ -59,73 +99,226 @@ func TestNamespaceSelectorHierarchicalFormat(t *testing.T) {
+	if err := nt.Watcher.WatchObject(kinds.ResourceGroup(),
+		configsync.RootSyncName,
+		configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.ResourceGroupHasObjects(selectedResourcesWithBookstoreNSSAndShoestoreNSS),
+			testpredicates.ResourceGroupHasNoObjects(unselectedResourcesWithBookstoreNSSAndShoestoreNSS),
+		}); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	validateSelectedAndUnselectedResources(nt,
+		selectedResourcesWithBookstoreNSSAndShoestoreNSS,
+		unselectedResourcesWithBookstoreNSSAndShoestoreNSS,
+	)
+
+	nt.Logger.Info("Update Namespace's label to make it unselected, resources should NOT be selected")
+	nt.MustMergePatch(bookstoreNamespace, `{"metadata":{"labels":{"app":"other"}}}`)
+
+	nt.Logger.Info("Watching the ResourceGroup object until unselected resources are removed from the inventory")
+	if err := nt.Watcher.WatchObject(kinds.ResourceGroup(),
+		configsync.RootSyncName,
+		configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.ResourceGroupHasObjects(selectedResourcesWithShoestoreNSSOnly),
+			testpredicates.ResourceGroupHasNoObjects(unselectedResourcesWithShoestoreNSSOnly),
+		}); err != nil {
+		nt.T.Fatal(err)
+	}
 	if err := nt.WatchForAllSyncs(); err != nil {
 		nt.T.Fatal(err)
 	}
-	if err := nt.Validate(bookstoreCM.Name, bookstoreNS, &corev1.ConfigMap{}); err != nil {
+	validateSelectedAndUnselectedResources(nt,
+		append(selectedResourcesWithShoestoreNSSOnly),
+		append(unselectedResourcesWithShoestoreNSSOnly),
+	)
+
+	nt.Logger.Info("Update Namespace's label to make it selected again, resources should be selected")
+	nt.MustMergePatch(bookstoreNamespace, fmt.Sprintf(`{"metadata":{"labels":{"app":"%s"}}}`, bookstoreNS))
+
+	if err := nt.Watcher.WatchObject(kinds.ResourceGroup(),
+		configsync.RootSyncName,
+		configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.ResourceGroupHasObjects(selectedResourcesWithBookstoreNSSAndShoestoreNSS),
+			testpredicates.ResourceGroupHasNoObjects(unselectedResourcesWithBookstoreNSSAndShoestoreNSS),
+		}); err != nil {
 		nt.T.Fatal(err)
 	}
-	if err := nt.Validate(bookstoreRQ.Name, bookstoreNS, &corev1.ResourceQuota{}, resourceQuotaHasHardPods(nt, "1")); err != nil {
+	if err := nt.WatchForAllSyncs(); err != nil {
 		nt.T.Fatal(err)
 	}
-	if err := nt.Validate(shoestoreCM.Name, shoestoreNS, &corev1.ConfigMap{}); err != nil {
+	validateSelectedAndUnselectedResources(nt,
+		selectedResourcesWithBookstoreNSSAndShoestoreNSS,
+		unselectedResourcesWithBookstoreNSSAndShoestoreNSS,
+	)
+
+	nt.Logger.Info("Stop the admission webhook so the Namespace can be deleted")
+	// If the webhook is running, the namespace DELETE call will fail because the
+	// underlying resources cannot be deleted due to the admission-webhook.
+	nomostest.StopWebhook(nt)
```

Comment:
Why does the webhook block deletion? It looks like the namespace is created out-of-band, so presumably should not have Config Sync management labels. Is it because the namespace contains managed resources?

Reply:
Correct. When deleting a namespace, resources in the namespace will also be deleted. But they're managed by Config Sync, the webhook will reject the deletion.

I have the `underlying resources cannot be deleted` comment. It seems not clear? Clarified.